Makefile.include

```diff
@@ -4,23 +4,43 @@
 all:
 
 # set undefined variables
-RIOTBASE ?= $(shell dirname "$(lastword $(MAKEFILE_LIST))")
-RIOTBASE := $(abspath $(RIOTBASE))
-
-CCACHE_BASEDIR := $(RIOTBASE)
-
-RIOTCPU ?= $(RIOTBASE)/cpu
-RIOTCPU := $(abspath $(RIOTCPU))
-
-RIOTBOARD ?= $(RIOTBASE)/boards
-RIOTBOARD := $(abspath $(RIOTBOARD))
-
-RIOTPKG ?= $(RIOTBASE)/pkg
+RIOTBASE       ?= $(shell dirname "$(lastword $(MAKEFILE_LIST))")
+CCACHE_BASEDIR ?= $(RIOTBASE)
+RIOTCPU        ?= $(RIOTBASE)/cpu
+RIOTBOARD      ?= $(RIOTBASE)/boards
+RIOTPKG        ?= $(RIOTBASE)/pkg
+RIOTPROJECT    ?= $(shell git rev-parse --show-toplevel 2>/dev/null || pwd)
+GITCACHE       ?= $(RIOTBASE)/dist/tools/git/git-cache
+APPDIR         ?= $(CURDIR)
+BINDIRBASE     ?= $(APPDIR)/bin
+BINDIR         ?= $(BINDIRBASE)/$(BOARD)
+
+__DIRECTORY_VARIABLES := RIOTBASE CCACHE_BASEDIR RIOTCPU RIOTBOARD RIOTPKG GITCACHE RIOTPROJECT APPDIR BINDIRBASE BINDIR
+
+# Ensure that all directories are set and don't contains spaces.
+ifneq (, $(filter-out 1, $(foreach v,${__DIRECTORY_VARIABLES},$(words ${${v}}))))
```

Comment:
for the current variables in `__DIRECTORY_VARIABLES` this may work, but if variables get added later that are allowed to be empty, than this will fail. Can you default to any string in case `${${v}}` is empty? `${${v:-qwantz}}`